lib/bolt/config.rb

```diff
@@ -121,57 +160,143 @@ def self.from_project(project, overrides = {})
                Bolt::Util.read_optional_yaml_hash(project.config_file, 'config')
              end
 
-      data = { filepath: project.config_file, data: conf }
-
-      data = load_defaults(project).push(data).select { |config| config[:data]&.any? }
+      data = load_defaults(project).push(
+        filepath: project.config_file,
+        data: conf,
+        warnings: []
+      )
 
       new(project, data, overrides)
     end
 
     def self.from_file(configfile, overrides = {})
       project = Bolt::Project.create_project(Pathname.new(configfile).expand_path.dirname)
+
       conf = if project.project_file == project.config_file
                project.data
              else
                Bolt::Util.read_yaml_hash(configfile, 'config')
              end
-      data = { filepath: project.config_file, data: conf }
-      data = load_defaults(project).push(data).select { |config| config[:data]&.any? }
+
+      data = load_defaults(project).push(
+        filepath: project.config_file,
+        data: conf,
+        warnings: []
+      )
 
       new(project, data, overrides)
     end
 
-    def self.load_defaults(project)
+    def self.system_path
       # Lazy-load expensive gem code
       require 'win32/dir' if Bolt::Util.windows?
 
-      # Don't load /etc/puppetlabs/bolt/bolt.yaml twice
-      confs = if project.path == Bolt::Project.system_path
-                []
-              else
-                system_path = Pathname.new(File.join(Bolt::Project.system_path, 'bolt.yaml'))
-                [{ filepath: system_path, data: Bolt::Util.read_optional_yaml_hash(system_path, 'config') }]
-              end
-
-      user_path = begin
-                    Pathname.new(File.expand_path(File.join('~', '.puppetlabs', 'etc', 'bolt', 'bolt.yaml')))
-                  rescue ArgumentError
-                    nil
-                  end
-
-      confs << { filepath: user_path, data: Bolt::Util.read_optional_yaml_hash(user_path, 'config') } if user_path
+      if Bolt::Util.windows?
+        Pathname.new(File.join(Dir::COMMON_APPDATA, 'PuppetLabs', 'bolt', 'etc'))
+      else
+        Pathname.new(File.join('/etc', 'puppetlabs', 'bolt'))
+      end
+    end
+
+    def self.user_path
+      Pathname.new(File.expand_path(File.join('~', '.puppetlabs', 'etc', 'bolt')))
+    rescue StandardError
+      nil
+    end
+
+    # Loads a 'bolt-defaults.yaml' file, which contains default configuration that applies to all
+    # projects. This file does not allow project-specific configuration such as 'hiera-config' and
+    # 'inventoryfile', and nests all default inventory configuration under an 'inventory-config' key.
+    def self.load_bolt_defaults_yaml(dir)
```

Comment:
I like having this as a separate function because it might be nice to keep these warnings when we drop `bolt.yaml`, and will make the changes to dropping that minimal since loading each file is already a separate codepath.